### ISSUE TITLE
Escape the adapter-specific string literal delimiter within the `string_literal` macro

### ISSUE THREAD

**Issue by dbeatty10** [dbt-labs/dbt-adapters#293]
In my point of view, the proposed workaround is a bug. So anyone using it, should stop and fix their code. Also, the macro `dbt.string_literal` should escape single quotes, so doubling like this :  `dbt.string_literal(dbt.escape_single_quotes(value))` should produce too many quotes, if it doesn't, one of those macros has a bug.

_Originally posted by @nicods-fr in https://github.com/dbt-labs/dbt-core/issues/10654#issuecomment-2331295325_

**Comment by james-johnston-thumbtack** [comment]
+1 for this... also I think the `string_literal` macro documentation at https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#string_literal should have a big fat warning explaining that the strings are NOT being escaped here.  I incorrectly assumed that this macro would escape strings, imagine my surprise when it does not!

**Comment by james-johnston-thumbtack** [comment]
There is another problem besides the single quotes.  Jinja strings can contain `\` characters which will be misinterpreted as an escape sequence, because `dbt.escape_single_quotes` does not handle that properly.  Now I _really_ can't think of any use case that `dbt.escape_single_quotes` would actually be useful for, since if I want to escape single quotes, I probably want to escape the character that starts an escape sequence as well.

As part of this issue, I think care should be taken so that _any_ Jinja string can be used.  The grammar for each SQL dialect should be very carefully consulted to ensure that everything that needs to be escaped will be escaped.

Example code that fails with BigQuery:
```
{% set json_document %}
{
    "null": null,
    "string": "Hello \"Bob\"!",
    "bool": true,
    "array": [2, 3, 1],
    "object": {
        "val_2": 100,
        "val_1": 300,
        "val_3": 200
    }
}
{% endset %}

SELECT JSON {{ dbt.string_literal(dbt.escape_single_quotes(json_document)) }} AS json_literal
```
It fails because the `\` character in the JSON string literal needs to be double-escaped as part of turning into a SQL string literal, i.e. `\\`.  I.e. the `escape_single_quotes` macro should have done a string replace of `\` to `\\`.

**Comment by james-johnston-thumbtack** [comment]
It turns out that the above issue breaks BigQuery unit tests as well: https://github.com/dbt-labs/dbt-adapters/blob/7a7346892415598c3b5901a76ba4ed9cd22084fd/dbt-adapters/src/dbt/include/global_project/macros/unit_test_sql/get_fixture_sql.sql#L87

because the unit test also fails to escape backslashes, which BigQuery requires.

**Comment by vergenzt** [comment]
Yes - I've been exasperated with this issue in dbt for YEARS. As someone with a software engineering background, injecting string literals as `'{{ my_string_here }}'` pains me so much.

I also cannot imagine why anyone would ever NOT want `dbt.string_literal(...)` to _escape_ the string safely.

**Comment by vergenzt** [comment]
The culture in the dbt ecosystem of not escaping strings when embedding is driving me crazy and it has for years. I just encountered it again in an adapter implementation: https://github.com/databricks/dbt-databricks/blob/8cda62ee19d01e0670e3156e652841e3ffd3ed41/dbt/include/databricks/macros/relations/components/comment.sql#L3

(My model description is long and had a possessive apostrophe in it. 😒)

We've got to fix this.

From the other issue:

> These macros originally came from the dbt_utils package, and they were ported over as-is. Since they've been around for a long time and many folks may be using them with full knowledge of their current implementations, we'd be unlikely to make that change to avoid breaking anyone.

There is some (minor) precedent for `dbt.<type>` producing a literal of that type (when the default string repr via Jinja doesn't work like it does for e.g. numbers): [`dbt.date(...)`](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#date). The exact example in the docs is `{{ dbt.date(2023, 10, 4) }}` producing a date literal (via a function call, but still).

The easiest way to avoid breaking backcompat is to use a new name for the new implementation.

I propose:

## 1. add macro `dbt.string(value)`
This macro should transform a Jinja string value into an identical value in the database's string literal representation.

Example implementation:
```jinja
{% macro postgres__string(value) -%}
  '{{ value | replace("'", "''") }}'
{%- endmacro %}
```

IMHO this macro should simply live in the ["String functions"](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#string-functions) section of the docs, and the ["String literal functions"](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#string-literal-functions) macros should be deprecated, and their documentation moved to the end of the ["String functions"](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#string-functions) section.

Adapter integration tests can be added which run `select {{ dbt.string('<complicated string test case>') }}` and check that the output cell value exactly matches the input test case string.

## 2. deprecate `dbt.escape_single_quotes` and `dbt.string_literal`

In favor of `dbt.string`. Explain the (security, robustness) advantages of the latter in a warning blurb at the tops of their docs descriptions.

## 3. profit

Start simplifying callers of the deprecated macros: https://github.com/search?q=org%3Adbt-labs+%2Fdbt%5C.%28string_literal%7Cescape_single_quotes%29%2F+path%3A*.sql&type=code

---

I would be delighted to help contribute this, but I'd love some feedback from maintainers on whether it would be welcome first.
